src/ophyd_async/core/_soft_signal_backend.py

```diff
@@ -1,244 +1,218 @@
-        except TypeError:
-            dk["dtype_numpy"] = ""  # type: ignore
-        return dk
-
-    def make_initial_value(self, datatype: type[T] | None) -> T:
-        if datatype is None:
-            return cast(T, None)
-
-        return datatype()
-
-
-class SoftArrayConverter(SoftConverter):
-    def get_datakey(self, source: str, value, **metadata) -> DataKey:
-        dtype_numpy = ""
-        if isinstance(value, list):
-            if len(value) > 0:
-                dtype_numpy = np.dtype(type(value[0])).descr[0][1]
-        else:
-            dtype_numpy = np.dtype(value.dtype).descr[0][1]
+class SequenceStrSoftConverter(SoftConverter[Sequence[str]]):
+    def write_value(self, value: Any) -> Sequence[str]:
+        return [str(v) for v in value] if value else []
 
-        return {
-            "source": source,
-            "dtype": "array",
-            "dtype_numpy": dtype_numpy,  # type: ignore
-            "shape": [len(value)],
-            **metadata,
-        }
 
-    def make_initial_value(self, datatype: type[T] | None) -> T:
-        if datatype is None:
-            return cast(T, None)
+@dataclass
+class SequenceEnumSoftConverter(SoftConverter[Sequence[EnumT]]):
+    datatype: type[EnumT]
 
-        if get_origin(datatype) == abc.Sequence:
-            return cast(T, [])
+    def write_value(self, value: Any) -> Sequence[EnumT]:
+        return [self.datatype(v) for v in value] if value else []
 
-        return cast(T, datatype(shape=0))  # type: ignore
 
+@dataclass
+class NDArraySoftConverter(SoftConverter[Array1D]):
+    datatype: np.dtype
 
-class SoftEnumConverter(SoftConverter):
-    choices: tuple[str, ...]
+    def write_value(self, value: Any) -> Array1D:
+        return np.array(value or (), dtype=self.datatype)
 
-    def __init__(self, datatype: RuntimeSubsetEnum | type[Enum]):
-        if issubclass(datatype, Enum):  # type: ignore
-            self.choices = tuple(v.value for v in datatype)
-        else:
-            self.choices = datatype.choices
 
-    def write_value(self, value: Enum | str) -> str:
-        return value  # type: ignore
+@dataclass
+class EnumSoftConverter(SoftConverter[EnumT]):
+    datatype: type[EnumT]
 
-    def get_datakey(self, source: str, value, **metadata) -> DataKey:
-        return {
-            "source": source,
-            "dtype": "string",
-            # type ignore until https://github.com/bluesky/event-model/issues/308
-            "dtype_numpy": "|S40",  # type: ignore
-            "shape": [],
-            "choices": self.choices,
-            **metadata,
-        }
-
-    def make_initial_value(self, datatype: type[T] | None) -> T:
-        if datatype is None:
-            return cast(T, None)
-
-        if issubclass(datatype, Enum):
-            return cast(T, list(datatype.__members__.values())[0])  # type: ignore
-        return cast(T, self.choices[0])
+    def write_value(self, value: Any) -> EnumT:
```

Comment:
In typing world then we will never `signal.put()` anything other than `None | SignalDatatypeT`. However in real life people could put anything to it, so I chose to make the `Converter` be true to this rather than the stricter typing sense. This could do with a comment...